Create 001-ClusterRepresentation.md (#26)
kroxylicious/design: proposals/001-ClusterRepresentation.md

A _virtual clusster_ references exactly one _physical cluster_. 

A _virtual cluster_ may reference a `filter chain`.  This provides zero or more filters that the RPCs will pass through before arriving at the physical cluster.

A _virtual cluster_ enumerates the virtual brokers that comprise the virtual cluster.

##### Virtual Brokers

Each broker has a reference to an endpoint that will provide the way in to the broker.   For the SNI case, there is also an addresses matcher.  It is an error
for more than one broker to share an endpoint unless the endpoint type is TLS and the SNI matching address is defined.

Each virtual broker references exactly one physical broker which must exist within the referenced physical cluster.  It is an error if there is not a
one-to-one correspondence between downstream and upstream brokers.

TLS key material may be provided either at the virtual cluster level or at the individual broker level.

```yaml
- name: my-public-cluster
  physicalClusterRef: my-private-cluster
  filterChainRef: my-upstream-filter-chain
  brokers:
  - name: broker-1
    endpointRef: my-tls-endpoint
    sniMatchAddress: broker-1-public.example.com

>     upstreamBrokerRef: broker-1

Reviewer: ```suggestion
    physicalBrokerRef: broker-1
```

I presume that we'll use the new terminology

    # or brokers themselves provide the key/cert
    - tls:
        key:
        cert:
  - name: broker-2
    ...
  # cluster wide TLS cert
  tls:
    key:
    cert:
```

#### Physical Clusters

> An _physical cluster_ is the in-model representation of a running kafka cluster.   The physical clusters list specifies all the kafka clusters that known to Kroxylicious.

Reviewer: ```suggestion
A _physical cluster_ is the in-model representation of a running kafka cluster.   The physical clusters list specifies all the kafka clusters that are known to Kroxylicious.
```

> An _physical cluster_ may reference a `filter chain`.  This provides zero or more *additional* filters that the RPCs will pass through before passing to the brokers

Reviewer: ```suggestion
A _physical cluster_ may reference a `filter chain`.  This provides zero or more *additional* filters that the RPCs will pass through before passing to the brokers
```

<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->

# Cluster Representation in the Model

This proposal introduces the concept of _cluster_ into the model.

> A _virtual cluster_ is a kafka cluster that the clients connect to.   The virtual cluster can be thought of as a 'virtual' cluster or a 'facade' to

Reviewer: ```suggestion
A _virtual cluster_ is a Kafka cluster that the clients connect to.  It can be thought of as a 'logical' cluster or a 'facade' to
```

Thinking either use the word logical here or just remove this part and make it `can be thought of as a facade..` .

> A _downstream cluster_ is a kafka cluster that the clients connect to.   The downstream cluster can be thought of as a 'virtual' cluster or a 'facade' to

Reviewer 2: Yeah. I'm much less sure about `physical` too. Toyed with `underlying` and `target` but `physical` worked well for as being the opposite of `virtual` and really drives the point home for me. 

As far as each Kroxy is concerned each upstream cluster is the `physical` one even if it happens to be another kroxy instance thats an implementation detail that doesn't concern the given instance (otherwise we have a leaky abstraction - but I can see arguments for co-operation between proxies...).

> A _downstream cluster_ is a kafka cluster that the clients connect to.   The downstream cluster can be thought of as a 'virtual' cluster or a 'facade' to

Reviewer: I struggled with the `upstream` and `downstream` terms as well, I think `physical` and `virtual` are much better replacements and are explicit in nature to spare the users any non-intuitive mental model mapping (e.g. "oh it mentions "cluster" only, so this must be the physical one, since the other one is called "virtual" ) 

I think either `virtual` and `physical` or `virtual` and `target` would be good alternatives

> A _downstream cluster_ is a kafka cluster that the clients connect to.   The downstream cluster can be thought of as a 'virtual' cluster or a 'facade' to

Author: I'll adopt virtual/physical nomenclature.

> A _downstream cluster_ is a kafka cluster that the clients connect to.   The downstream cluster can be thought of as a 'virtual' cluster or a 'facade' to

Author: Done

one or more _upstream clusters_.    An _upstream cluster_ represents the physical kafka cluster.  There is a one to one correspondance between
a _upstream cluster_ and a physical cluster in the world.

> There will be a mechansim to route flexibly between downstreams and upstreams.  This will allow useful topologies to built to serve different use-cases.

Reviewer: ```suggestion
There will be a mechansim to route flexibly between downstreams and upstreams.  This will allow building useful topologies to serve different use-cases.
```

- *one to one* - the simple exposure of a physical cluster 
- *many to one* - kroxy presents many downstream clusters which route to a single upstream - this would support a multitenant use-cases 

(There's a possiblity of *one-to-many* which a single downstream routes to several upstreams as if they were one, which might be useful for presenting several

> kafka clusters as if they were one, however, supporting transactions across two clusters would be difficult.  We consider won't this use-case futrther).

Reviewer: ```suggestion
kafka clusters as if they were one, however, supporting transactions across two clusters would be difficult.  We won't consider this use-case further).
```

> kafka clusters as if they were one, however, supporting transactions across two clusters would be difficult.  We consider won't this use-case futrther).

Author: Done

## Current situation

Currently Kroxyilicious binds a single port to the downstream and connects to a single upstream broker.  There is no support for clusters comprised of more
then one broker.

## Motivation

1. Ability for Kroxylicious to expose multiple physical kafka clusters.
1. Ability for Kroxylicious to support clusters formed of more than one broker.
1. Ability for Kroxylicious to provide a building block to support multi-tenancy.

## Proposal

The proposal will change some existing concepts and introduce some new ones.  These are discussed first.  Then, we talk about the changes in high level responsibilities.

### Key Objects

![alternative text](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.github.com/kroxylicious/design/001-ClusterRepresentation/proposals/001/UML/uml.puml)

#### Endpoints

Currently `KafkaProxy` uses the `ProxyConfig` object to get a single address to bind to.  

* For the non-TLS case, multiple endpoints are required to support >1 brokers. 

> * For TLS case, SNI gives a mechanism to route the traffic to a >1 brokers even when a single endpoint is used, however there may be performance or operational reasons to favour the ability to define endpoints.

Reviewer: ```suggestion
* For TLS case, SNI gives a mechanism to route the traffic to >1 brokers even when a single endpoint is used, however there may be performance or operational reasons to favour the ability to define endpoints.
```

`ProxyConfig` will be replaced by an list of _Endpoint_ objects.  

```
- name: endpoint name
  address: bind address/port
  type: PLAIN|TLS
```

#### Downstream Clusters

Conceptually the _downstream cluster_ represents the cluster that the kafka client connects to.  The downstream clusters list specifies
all the kafka clustsers are being presented by Kroxylicious. 

A _downstream clustser_ references exactly one _upstream cluster_. 

A _downstream cluster_ may reference a `filter chain`.  This provides zero or more filters that the RPCs will pass through before arriving at the upstream cluster.

> A _downstream cluster_ enumerates the downstream brokers that comprise the downstream cluster.

Reviewer 3: Could we add something here explicitly stating the relationship between Downstream Cluster and Endpoint?

Something like `A downstream cluster is associated with a single Endpoint. Multiple downstream clusters may refer to the same Endpoint`

Same with TLS like `A downstream cluster can provide TLS configuration which is used for XYZ`

##### Downstream Brokers

Each broker has a reference to an endpoint that will provide the way in to the broker.   For the SNI case, there is also an addresses matcher.  It is an error
for more than one broker to share an endpoint unless the endpoint type is TLS and the SNI matching address is defined.

Each downstream broker references exactly one upstream broker which must exist within the referenced upstream cluster.  It is an error if there is not a

> one-to-one correspondence between downstream and upstream brokers.

Reviewer 2: Does this hold true in the multi tenant use case? @robobario and I are imagining providing N `virtual` clusters compose of `3` brokers each being serviced by a **single** `physical` cluster composed of `6` brokers. 

We could establish fixed mappings from `virtual` brokers to `physical` brokers initially but when we use cruse control to rebalance topics across the `6 physical` brokers those fixed mapping no longer hold.  As the partitions could be moved to the brokers not included in the one to one mapping. We potentially need to be able to map multiple `physical` brokers to a single `virtual` broker. 

We may not need to support this the MVP but any design we come up with should be open to extending in this direction. In the MVP we could enforce that all `virtual` clusters have the same broker counts as the `physical` clusters.

Author: > Does this hold true in the multi tenant use case? @robobario and I are imagining providing N `virtual` clusters compose of `3` brokers each being serviced by a **single** `physical` cluster composed of `6` brokers.

Can that work?   I had been working under the (unspoken) assumption that a virtual cluster had n virtual brokers where n matched the number of the physical brokers on the cluster. It is possible I'm the outlier here.

> We could establish fixed mappings from virtual brokers to physical brokers initially but when we use cruise control to rebalance topics across the 6 physical brokers those fixed mapping no longer hold.

Exactly.

> We potentially need to be able to map multiple physical brokers to a single virtual broker.

I'm not sure that could work.  When a topic is created, the kafka client decides which broker each partition/replicas go.   If there's not a one to one mapping between virtual broker and physical ones I think this would break down.

It might be possible to have kroxy expose only those brokers hosting partitions where the tenant has partitions but this does seen to bring a fair amount of extra complexity.

What's the advantage for exposing only a subset of the physical brokers?

Reviewer 2: On reflection I don't think `N physical` brokers to `1 virtual` broker doesn't work and is unnecessary but I think we do need to support a subset of brokers (at some point).

The use-case I am thinking of is something like offering small clusters on a much larger shared cluster, in the tens or hundreds of brokers. I don't see it making any realistic sense to expose hundreds of brokers through the proxy to a single `virtual` cluster. If we expand to that point having a static map of `virtual` clusters to `physical` brokers becomes very difficult to manage. 

For the record I'm not saying we need to fix this during MVP, just that we need to keep this end goal in mind and keep options open that make that support easier.

Author: > For the record I'm not saying we need to fix this during MVP, just that we need to keep this end goal in mind and keep options open that make that support easier.
+1

I'll call that out as a goal.

Author: Done.

> TLS key material may be provided either at the downstream cluster level or at the individual broker level.

Reviewer 3: What would the cluster-level TLS configuration be used for? We know the broker SNI hostnames, but would we know the bootstrap hostname to apply it to? Should a public hostname be part of the downstream cluster configuration?

Author: It is for the certificate per cluster use-case. The cert's SAN would need to list the FQDN of each broker within the cluster.

Reviewer 3: Okay, I was more thinking of if we have a separate bootstrap host like:
`my-kafka-a.b.c.robeyoun-kafka.public.host`
and then per-broker hosts like `broker-1.my-kafka-a.b.c.robeyoun-kafka.public.host`

Where would we know how to map the bootstrap host to the downstream cluster SSL configuration?

Reviewer 3: edit: maybe we can model bootstrap as a downstream and upstream broker.

```yaml
- name: my-public-cluster
  upstreamClusterRef: my-private-cluster
  filterChainRef: my-upstream-filter-chain
  brokers:
  - name: broker-1
    endpointRef: my-tls-endpoint
    sniMatchAddress: broker-1-public.example.com

>     upstreamBrokerRef: broker-1

Author: Yes, do you see a problem with that assumption?

>     upstreamBrokerRef: broker-1

Reviewer 2: I think we've addressed the underlying challenges with this else where.

    # or brokers themselves provide the key/cert
    - tls:
        key:
        cert:
  - name: broker-2
    ...
  # cluster wide TLS cert
  tls:
    key:
    cert:
```

#### Upstream Clusters

An _upstream cluster_ represents a physical kafka cluster.   The upstream clusters list specifies all the kafka clustsers that are being exposed through Kroxylicious.  

> An _upstream cluster_ may reference a `filter chain`.  This provides zero or more *additional* filters that the RPCs will pass through before passing to the brokers

Author: yes, that matches my mental model.

of the physical cluster.  

An _upstream cluster_ enumerates the upstream brokers that comprise the physical cluster.

#### Upstream Brokers

Each upstream broker specifies the address of the physical broker and whether TLS is to be used.  It will also specify trust settings.

TLS trust material may be provided at the upstream broker level or at the individual broker level.

```
- name: my-private-cluster
  filterChainRef: my-downstream-filter-chain
  tls:
    trust:
    ...
  brokers:
  - name: broker-1
    address: broker-1-private:9092
    ...
    tls:
      trust:
      ...
  - name: broker-2
    address: broker-1-private:9092
    ...
    tls:
      trust:
```

#### Filter Chain

> Currently kroxylicious provide a single filter chain.  It will refactored so that we can support many filter chains, each identified by name. The `downstream cluster` and `upstream cluster` may reference a filter chain. 

Reviewer 2: I'm starting to think it might be better to express this as the `virtual` cluster defines a filter chain (potentially applied by the `frontendHandler` and the `physical` cluster defines a filter chain (again applied by the `backendHandler`). Those `virtual` and `physical` chain definitions could potentially be shared between multiples of each cluster type but there is no single over arching chain definition.

Author: Yes, maybe.  if each virtual cluster/cluster declares the filter chain in its own right, there's no problem passing cluster specific configuration.

```yaml
filterChains:
- name: my-upstream-filter-chain
  filters:
  - type: ApiVersions
  - type: BrokerAddress

```

### Responsibilities

All specified endpoints will be bound.

When a connection is made to an endpoint,  the system must resolve that connection to a downstream broker and hence a downstream cluster.

To do this, it resolves the endpoint and any SNI information against the model.   This should yield exactly one broker belonging to a downstream cluster.
It is an error otherwise and the connnection must be closed.

> If TLS is in use, the SSLContext can be generated from the downstream cluster definition.  This will be passed to Netty to let it complete the TLS handshake.

Reviewer 2: This feels very implementation specific compared to the other responsibilities. 

I see it as an important point to

The downstream broker and downstream cluster is used to identify the upstream broker and upstream cluster.

The upstream cluster and downstream broker provide the filter chain.

The handler chain connects to the upstream broker.

> The system must reload chnages to the model dynammically, without dropping established connections.

Author: > I think we should make it explicit that the reloaded config is applied by the existing connections.

yeah - I was pondering whether reconfiguring an upstream or downstream broker should close the established connections.  similarly for reconfiguring an upstream or downstream cluster should that close all the established connections of all the brokers?

I'm not sure how important it is to consider right now.  I think they might all be edge cases or require no action.

1. I don't think a use-case where a downstream cluster is pointed at a different upstream cluster is likely to materialise in the short term (tar pitting might be a use-case)
2. TLS cert renewal.  I don't think there is an necessity to disconnect existing connections.
3. Upstream cluster topology changes
- fewer brokers - connections will die as the physical upstream broker is stopped.
- more brokers - kafka's metadata discovery will handle this. No advantage gained by interrupting the connections to the existing brokers.

### Broker Address Filter

How will broker address filter map the RPCs that contain the upstream broker address to the downstream broker addresss that are resolveable to the client?

In the case where kroxy is being placed in front of a kafka cluster spanning a three-AZ cluster, the kroxylicious instances in the AZ won't need configuration to
*connect* to brokers in the other AZs.  However, BrokerAddressFilter will be to be capable of rewriting the the broker address for the whole cluster.
This leads us to the conclusion that Broker Address filter need separate configuration independent of that what could be derived from the upstream/downstream cluster mapping.

> TODO: Maybe used a regexp based mapping would be sufficient

Author: Yes I believe a mapping will work.  The mapping will need the port numbers.  There's the possibility that downstream and upstream port numbers would be different.

I've been thinking about where this mapping lives.   If it is exists as filter configuration, then in the multitenant case, we'd need a separate filter configuration for each tenant.  Whilst that would work, I'm wondering whether the cluster configuration should have have a map of arbitrary configuration that is made available to the filters at runtime (somehow).  

We could use this to pass extra information to the filters like the MultiTenantFilter (a tenant key) and BrokerAddressFilter (a mapping).

```yaml
- name: tenant1-cluster
  upstreamClusterRef: my-private-cluster
  filterChainRef: my-upstream-filter-chain
  filterVariables:
      io.kroxylicious.tenantKey: xyz4a
      io.kroxylicious.brokerAddressMapping: 
        broker-1.azA.internal.svc:1234: broker-A.myvanitydomain.com:9080
```

I think some of the variable name would be 'well-known' (we reserve names variable names beginning with `io.kroxylicious`.

> TODO: Maybe used a regexp based mapping would be sufficient

Reviewer 2: I was assuming we would generate the mapping table from the configuration rather than listing it explicitly as we are currently talking about having exhaustive mappings. 

But yeah maybe including it in the config is just simpler and easier to deal with for now.

> TODO: Maybe used a regexp based mapping would be sufficient

Author: > I was assuming we would generate the mapping table from the configuration rather than listing it explicitly as we are currently talking about having exhaustive mappings.

That was my initial thinking too.  I don't think that works. 

- Firstly, in the case where you have an AZ specific kroxy deployment, the mapping needs to know about the brokers on the other AZs.
- Secondly, there's no guarantee that the endpoint address/port that kroxy needs to bind is the same as the address/port that actually needs to be advertised to the client.

> TODO: Maybe used a regexp based mapping would be sufficient

Author: I'm now passing separate configuration to BrokerAddressFilter

  type: PLAIN|TLS
```

#### Virtual Clusters

Conceptually the _virtual cluster_ represents the cluster that the kafka client connects to.  The virtual clusters list specifies
all the kafka clusters are being presented by Kroxylicious. 

A _virtual clusster_ references exactly one _physical cluster_. 

A _virtual cluster_ may reference a `filter chain`.  This provides zero or more filters that the RPCs will pass through before arriving at the physical cluster.

A _virtual cluster_ enumerates the virtual brokers that comprise the virtual cluster.

##### Virtual Brokers

Each broker has a reference to an endpoint that will provide the way in to the broker.   For the SNI case, there is also an addresses matcher.  It is an error
for more than one broker to share an endpoint unless the endpoint type is TLS and the SNI matching address is defined.

Each virtual broker references exactly one physical broker which must exist within the referenced physical cluster.  It is an error if there is not a
one-to-one correspondence between downstream and upstream brokers.

TLS key material may be provided either at the virtual cluster level or at the individual broker level.

```yaml

> - name: my-public-cluster

Reviewer: I found it confusing when the yaml didn't include the `virtualClusters` key. I think it's more clear if it's something like this:
```yaml
...
virtualClusters:
- name: my-public-cluster
  upstreamClusterRef: my-private-cluster
  filterChainRef: my-downstream--filter-chain
  tls:
    key:
    cert:
  brokers:
 name: my-public-cluster
  upstreamClusterRef: my-private-cluster
  filterChainRef: my-upstream-filter-chain
  brokers:
  - name: broker-1
    endpointRef: my-tls-endpoint
    sniMatchAddress: broker-1-public.example.com
    physicalBrokerRef: broker-1
    # or brokers themselves provide the key/cert
    - tls:
        key:
        cert:
  - name: broker-2
    ...
  # cluster wide TLS cert
  tls:
    key:
    cert:
    ...
```

  physicalClusterRef: my-private-cluster
  filterChainRef: my-upstream-filter-chain
  brokers:
  - name: broker-1
    endpointRef: my-tls-endpoint
    sniMatchAddress: broker-1-public.example.com
    upstreamBrokerRef: broker-1
    # or brokers themselves provide the key/cert
    - tls:
        key:
        cert:
  - name: broker-2
    ...
  # cluster wide TLS cert
  tls:
    key:
    cert:
```

#### Physical Clusters

An _physical cluster_ is the in-model representation of a running kafka cluster.   The physical clusters list specifies all the kafka clusters that known to Kroxylicious.

An _physical cluster_ may reference a `filter chain`.  This provides zero or more *additional* filters that the RPCs will pass through before passing to the brokers
of the physical cluster.  

> An _physical cluster_ enumerates the brokers that comprise the physical cluster.

Reviewer: ```suggestion
A _physical cluster_ enumerates the brokers that comprise the physical cluster.
```

#### Physical Brokers

Each physical broker specifies the address of a broker and whether TLS is to be used.  It will also specify trust settings.

TLS trust material may be provided at the upstream broker level or at the individual broker level.

```yaml

> - name: my-private-cluster

Reviewer: same comment as above about including the `physicalClusters:` key

    address: broker-1-private:9092
    ...
    tls:
      trust:
      ...
  - name: broker-2
    address: broker-1-private:9092
    ...
    tls:
      trust:
```

#### Filter Chain

Currently kroxylicious provide a single filter chain.  It will refactored so that we can support many filter chains, each identified by name. The `virtual cluster` and `physical cluster` may reference a filter chain. 

```yaml
filterChains:
- name: my-upstream-filter-chain
  filters:
  - type: ApiVersions
  - type: BrokerAddress

```

### Responsibilities

All specified endpoints will be bound.

When a connection is made to an endpoint,  the system must resolve that connection to a virtual broker and hence a virtual cluster.

To do this, it resolves the endpoint and any SNI information against the model.   This should yield exactly one broker belonging to a virtual cluster.
It is an error otherwise and the connnection must be closed.

If TLS is in use, the SSLContext can be generated from the virtual cluster definition.  This will be passed to Netty to let it complete the TLS handshake.

The virtual broker and virtual cluster is used to identify the physical broker and physical cluster.

> The physical cluster and downstream broker provide the filter chain.

Reviewer: ```suggestion
The physical cluster and physical broker provide the filter chain.
```

The handler chain connects to the upstream broker.

The system must reload chnages to the model dynammically, without dropping established connections.

### Broker Address Filter

> How will broker address filter map the RPCs that contain the upstream broker address to the downstream broker addresss that are resolveable to the client?

Reviewer: ```suggestion
How will broker address filter map the RPCs that contain the physical broker address to the virtual broker addresss that are resolveable to the client?
```

In the case where kroxy is being placed in front of a kafka cluster spanning a three-AZ cluster, the kroxylicious instances in the AZ won't need configuration to

> *connect* to brokers in the other AZs.  However, BrokerAddressFilter will be to be capable of rewriting the the broker address for the whole cluster.

Reviewer: ```suggestion
*connect* to brokers in the other AZs.  However, BrokerAddressFilter will have to be capable of rewriting the broker address for the whole cluster.
```

      trust:
```

### High Level Responsibilities

#### Connection Handling

All declared endpoints will be bound.

When a connection is made to an endpoint,  the system must resolve that connection to a virtual broker and hence a virtual cluster.

To do this, it resolves the endpoint (and any SNI information) against the model.   This should yield exactly one broker belonging to a virtual cluster.
It is an error otherwise and the connnection must be closed.

If TLS is in use, the SSLContext can be generated from the virtual cluster definition.  This will be passed to Netty to let it complete the TLS handshake.

The virtual broker and virtual cluster is used to identify the physical broker and physical cluster.

The filter chains are constructed.

The handler chain connects to the physical broker.

#### Model Changes

The system must reload chnages to the model dynammically, without dropping established connections.

#### Broker Address Filter

The BrokerAddressFilter must map the RPC reponses that contain the physical broker address to the virtual broker addresses that are resolveable to the client.  DescribeCluster response is an example of an RPC that needs to be mapped.

> In the case where Kroxylicious is being placed in front of a kafka cluster spanning a three-AZ cluster, the kroxylicious instances in the AZ won't need configuration to *connect* to brokers in the other AZs.  However, BrokerAddressFilter must be capable of rewriting the the broker address for the whole cluster.

Reviewer: ```suggestion
In the case where Kroxylicious is being placed in front of a kafka cluster spanning a three-AZ cluster, the kroxylicious instances in the AZ won't need configuration to *connect* to brokers in the other AZs.  However, BrokerAddressFilter must be capable of rewriting the broker address for the whole cluster.
```

This leads us to the conclusion that Broker Address filter need separate configuration, independent of that what could be derived from the virtual/physical cluster mapping.

> The BrokerAddressFilter will accept a mapping, which will map the upstream broker addresses to routable addresses that correspond to the virtual brokers of the

Reviewer: ```suggestion
The BrokerAddressFilter will accept a mapping, which will map the physical broker addresses to routable addresses that correspond to the virtual brokers of the
```

cluster in question.

To do this, we can use the existing filter configuration mechanism.

```
  - type: BrokerAddress
    config:
      brokerAddressMapping:
        broker-1.azA.internal.svc:19080: broker-A.myvanitydomain.com:9092
        broker-2.azB.internal.svc:19080: broker-B.myvanitydomain.com:9092
        broker-3.azC.internal.svc:19080: broker-C.myvanitydomain.com:9092
```       

### Example Config Files

#### One to one case.

> # Single Cluster with 3 broker cluster, exposed via kroxy

Reviewer: ```suggestion
# Single Cluster with 3 brokers, exposed via kroxylicious
```

This leads us to the conclusion that Broker Address filter need separate configuration, independent of that what could be derived from the virtual/physical cluster mapping.

The BrokerAddressFilter will accept a mapping, which will map the upstream broker addresses to routable addresses that correspond to the virtual brokers of the
cluster in question.

To do this, we can use the existing filter configuration mechanism.

```
  - type: BrokerAddress
    config:
      brokerAddressMapping:
        broker-1.azA.internal.svc:19080: broker-A.myvanitydomain.com:9092
        broker-2.azB.internal.svc:19080: broker-B.myvanitydomain.com:9092
        broker-3.azC.internal.svc:19080: broker-C.myvanitydomain.com:9092
```       

### Example Config Files

#### One to one case.

# Single Cluster with 3 broker cluster, exposed via kroxy

```yaml
endpoints:
- name: my-tls-endpoint
  type: TLS
  bindAddress: 0.0.0.0:9092
virtualClusters:
- name: my-public-cluster

>   upstreamClusterRef: my-private-cluster

Reviewer: ```suggestion
  physicalClusterRef: my-private-cluster
```

    address: broker-3.azC.internal.svc:19080
```

#### Many-to-one case (multi tenancy)

# Single cluster, exposed to two tenants.

Note that in this case we've define two separate virtual clusters, one for each tenant.  We leveraged the fact that filters can be define at the
virtual cluster level to pass in tenant specific information to the filter. Specifically, this is how we pass in the tenant key (prefix) and the
broker address mapping.

```yaml
endpoints:
- name: my-tls-endpoint
  type: TLS
  bindAddress: 0.0.0.0:9092
virtualClusters:
- name: my-tenant1
  physicalClusterRef: my-big-cluster
  tls:
    key:
    cert:
  filters:
  - type: ApiVersions
  - type: BrokerAddress
    config:
      brokerAddressMapping:
        broker-1.azA.internal.svc:19080: broker-1.my-tenant1.kafka.com:9092
        broker-2.azB.internal.svc:19080: broker-2.my-tenant1.kafka.com:9092
        broker-3.azC.internal.svc:19080: broker-3.my-tenant1.kafka.com:9092
  - type: MultiTenantFilter
    config:
      tenantKey: z4de # prefix used to prefix objects owned by the tenant.
  brokers:
  - name: broker-1
    sniMatchAddress: broker-1.my-tenant1.kafka.com
    endpointRef: my-tls-endpoint

>     upstreamBrokerRef: private-broker-1

Reviewer: ```suggestion
    physicalBrokerRef: private-broker-1
```

  - name: broker-2
    sniMatchAddress: broker-2.my-tenant1.kafka.com
    endpointRef: my-tls-endpoint

>     upstreamBrokerRef: private-broker-2

Reviewer: ```suggestion
    physicalBrokerRef: private-broker-2
```

  - name: broker-3
    sniMatchAddress: broker-3.my-tenant1.kafka.com
    endpointRef: my-tls-endpoint

>     upstreamBrokerRef: private-broker-3

Reviewer: ```suggestion
    physicalBrokerRef: private-broker-3
```

- name: my-tenant2

>   upstreamClusterRef: my-big-cluster

Reviewer: ```suggestion
  physicalClusterRef: my-big-cluster
```

  tls:
    key:
    cert:
  filters:
  - type: ApiVersions
  - type: BrokerAddress
    config:
      brokerAddressMapping:
        broker-1.azA.internal.svc:19080: broker-1.my-tenant2.kafka.com:9092
        broker-2.azB.internal.svc:19080: broker-2.my-tenant2.kafka.com:9092
        broker-3.azC.internal.svc:19080: broker-3.my-tenant2.kafka.com:9092
  - type: MultiTenantFilter
    config:
      tenantKey: ab7ge
  brokers:
  - name: broker-1
    sniMatchAddress: broker-1.my-tenant2.kafka.com
    endpointRef: my-tls-my-tls-endpoint

>     upstreamBrokerRef: private-broker-1

Reviewer: ```suggestion
    physicalBrokerRef: private-broker-1
```

  - name: broker-2
    sniMatchAddress: broker-2.my-tenant2.kafka.com
    endpointRef: my-tls-endpoint

>     upstreamBrokerRef: private-broker-2

Reviewer: ```suggestion
    physicalBrokerRef: private-broker-2
```

  - name: broker-3
    sniMatchAddress: broker-3.my-tenant2.kafka.com
    endpointRef: my-tls-endpoint

>     upstreamBrokerRef: private-broker-3

Reviewer: ```suggestion
    physicalBrokerRef: private-broker-3
```

## Current situation

Kroxyliciousis currently limited to exposing a single broker of a single cluster.

## Proposal

The proposal will change some existing concepts and introduce some new ones.  These are discussed first.  Then, we talk about the changes in high level responsibilities.

### Key Objects

![alternative text](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.github.com/kroxylicious/design/001-ClusterRepresentation/proposals/001/UML/uml.puml)

#### Endpoints

Currently `KafkaProxy` uses the `ProxyConfig` object to get a single address to bind to.  

* For the non-TLS case, multiple endpoints are required to support >1 brokers. 
* For TLS case, SNI gives a mechanism to route the traffic to >1 brokers even when a single endpoint is used, however there may be performance or operational reasons to favour the ability to define endpoints.

> `ProxyConfig` will be replaced by an list of _Endpoint_ objects.  

Author: I was thinking the endpoint specify if it was TLS or not. It would probably also be where we configure cipher suites etc though I haven't specified this.  If we pushed the endpoint configuration into the upstream, we'd need to resolve any conflicts. I think having named endpoints is simpler to avoid those issues.

> `ProxyConfig` will be replaced by an list of _Endpoint_ objects.  

Reviewer 2: Yeah ok. Lets leave them separated.

```yaml
- name: endpoint name
  address: bind address/port
  type: PLAIN|TLS
```

#### Virtual Clusters

Conceptually the _virtual cluster_ represents the cluster that the kafka client connects to.  The virtual clusters list specifies
all the kafka clusters are being presented by Kroxylicious. 

A _virtual cluster_ references exactly one _physical cluster_. 

A _virtual cluster_ may define a chain of filters.  This provides zero or more filters that the RPCs will pass through as they traverse the virtual cluster,
before they reach the physical cluster.

A _virtual cluster_ enumerates the virtual brokers that comprise the virtual cluster.

##### Virtual Brokers

Each broker has a reference to an endpoint that will provide the way in to the broker.   For the SNI case, there is also an addresses matcher.  It is an error
for more than one broker to share an endpoint unless the endpoint type is TLS and the SNI matching address is defined.

Each virtual broker references exactly one physical broker which must exist within the referenced physical cluster.  It is an error if there is not a

> one-to-one correspondence between physical and virtual brokers.

Author: By physical broker here I am referring to the model object, rather than the the actual Kafka Cluster. Am I trying to say that the brokers defined by virtual cluster configuration needs to match the brokers the defined by the physical cluster model object.

For a scaling event, after scaling up the number of brokers on the Apache Kafka Cluster, an operator would update both the virtual cluster definition and the physical cluster definition consumed by Kroxy in lock step.

At the heart of this proposal is that the user of kroxy configures kroxy to know about the target cluster topology (i.e. static configuration enumerating the target cluster's brokers) and that configuration can be updated at runtime.    An alternative (not explored by this proposal) is let kroxy discover the broker topology of the target cluster.   Kroxy configuration would just need a bootstrap url of the target cluster.  If we took this option, we would need to require TLS connections at the downstream (client side).  This is so that SNI could be used to distinguish the broker traffic arriving at a single endpoint..

```yaml
virtualClusters:
- name: my-public-cluster
  physicalClusterRef: my-private-cluster
  filters:
  - type: Filter1
    config:
       foo: bar
  - type: Filter2
  brokers:
  - name: broker-1
    endpointRef: my-tls-endpoint
    sniMatchAddress: broker-1-public.example.com
    physicalBrokerRef: broker-1
    # or brokers themselves provide the key/cert
    - tls:
        key:
        cert:
  - name: broker-2
    ...
  # cluster wide TLS cert
  tls:
    key:
    cert:
```

#### Physical Clusters

An _physical cluster_ is the in-model representation of a running kafka cluster.   The physical clusters list specifies all the kafka clusters that known to Kroxylicious.

A physical cluster_ may define a chain of filters.  This provides zero or more filters that the RPCs will pass through as they traverse the physical cluster.
(Note that it is possible for both a filter chain to be define at both the virtual cluster and physical cluster level.  RPCs pass through both sets of filters
in the order they are defined).

An _physical cluster_ enumerates the brokers that comprise the physical cluster.

#### Physical Brokers

> Each physical broker specifies the address of a broker and whether TLS is to be used.  It will also specify trust settings.

Author: I don't understand how that could work.

The virtual brokers needs to be in charge of the advertised address.   For the many-to-one case, the my broker1 FQDN needs to be different to your broker1 FQDN.

TLS trust material may be provided at the virtual broker level or at the individual broker level.

```yaml
physicalClusters:
- name: my-private-cluster
  filters:
  - type: Filter1
    config:
       foo: bar
  - type: Filter2
  tls:
    trust:
    ...
  brokers:
  - name: broker-1

>     address: broker-1-private:9092

Author: Like I said above I don't see how having advertisedAddress as an attribute of the physical broker helps.

#### Model Changes

The system must reload chnages to the model dynammically, without dropping established connections.

#### Broker Address Filter

The BrokerAddressFilter must map the RPC reponses that contain the physical broker address to the virtual broker addresses that are resolveable to the client.  DescribeCluster response is an example of an RPC that needs to be mapped.

In the case where Kroxylicious is being placed in front of a kafka cluster spanning a three-AZ cluster, the kroxylicious instances in the AZ won't need configuration to *connect* to brokers in the other AZs.  However, BrokerAddressFilter must be capable of rewriting the broker address for the whole cluster.
This leads us to the conclusion that Broker Address filter need separate configuration, independent of that what could be derived from the virtual/physical cluster mapping.

The BrokerAddressFilter will accept a mapping, which will map the addresses returned by the upstream broker to routable addresses that route to the virtual brokers of the cluster in question.

To do this, we can use the existing filter configuration mechanism.

```
  - type: BrokerAddress
    config:

>       brokerAddressMapping:

Reviewer 2: If we add the `advertisedAddress` to the `physical broker` definition we already have the relationship established and I worry about having to generate the mappings twice in the config file. 

We could then provide the `brokerAddressMapping` as a derived property of the `virtual cluster` configuration when injected into the filters. However that implies the `virtual config` stops being a dumb value object and starts being a right domain object.

As an aside this thought was leading down the track of wondering about how to do Inversion Of Control (IOC) or Dependency Injection (DI) with filter instances. However I think making the derived address mapping a property of the `virtual cluster` config side steps that issue and leaves it in the long grass.
